File: backend/app/models/session.py
"""Session models for the AI Education Platform."""
import json
from datetime import datetime
from typing import TYPE_CHECKING
from uuid import UUID, uuid4

from sqlmodel import Field, Relationship, SQLModel

if TYPE_CHECKING:
    from .child import Child
    from .lesson import Lesson

# TODO[BE]: Add validation for session duration limits
# TODO[DB]: Implement session cleanup for expired sessions
# SECURITY[AUTH]: Add session token validation


class ChatMessage(SQLModel, table=True):
    """Chat message model for session conversations."""

    __tablename__ = "chat_messages"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    session_id: UUID = Field(foreign_key="sessions.id", index=True)
    role: str  # "child", "agent"
    content: str
    content_type: str = Field(default="text")  # "text", "audio"
    audio_url: str | None = Field(default=None)
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    message_metadata: str = Field(default="{}")  # JSON metadata

    # Content safety
    is_flagged: bool = Field(default=False)
    flagged_reason: str | None = Field(default=None)

    # Relationships
    session: "Session" = Relationship(back_populates="messages")


class Session(SQLModel, table=True):
    """Learning session model for tracking child-lesson interactions."""

    __tablename__ = "sessions"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    child_id: UUID = Field(foreign_key="children.id", index=True)
    lesson_id: UUID = Field(foreign_key="lessons.id", index=True)
    subject: str  # "arabic", "english", "islamic"
    agent_id: str

    # Session state
    status: str = Field(default="active")  # "active", "paused", "completed", "abandoned"
    start_time: datetime = Field(default_factory=datetime.utcnow)
    end_time: datetime | None = Field(default=None)

    # Progress tracking
    activities_completed: str = Field(default="[]")  # JSON array
    current_activity: str | None = Field(default=None)
    score: int = Field(default=0)
    time_spent: int = Field(default=0)  # minutes
    engagement_level: str | None = Field(default=None)  # "low", "medium", "high"
    hints_used: int = Field(default=0)

    # Assessment
    final_score: int | None = Field(default=None)
    points_earned: int = Field(default=0)
    badges_earned: str = Field(default="[]")  # JSON array

    # Metadata
    device_info: str = Field(default="{}")  # JSON

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime | None = Field(default=None)

    # Relationships
    child: "Child" = Relationship(back_populates="sessions")
    lesson: "Lesson" = Relationship(back_populates="sessions")
    messages: list[ChatMessage] = Relationship(back_populates="session")

    @property
    def activities_completed_list(self) -> list[str]:
        """Parse activities_completed JSON field into a list."""
        return json.loads(self.activities_completed)

    @property
    def badges_earned_list(self) -> list[str]:
        """Parse badges_earned JSON field into a list."""
        return json.loads(self.badges_earned)

    class Config:
        """SQLModel configuration."""

        validate_assignment = True

File: packages/schemas/src/session.ts
import { z } from "zod";
import { SubjectEnum, TimestampSchema } from "./common";

// Chat message schema
export const ChatMessageSchema = z.object({
  id: z.string().uuid(),
  role: z.enum(["child", "agent"]),
  content: z.string(),
  contentType: z.enum(["text", "audio"]).default("text"),
  audioUrl: z.string().url().optional(),
  timestamp: z.string().datetime(),
  metadata: z.record(z.any()).optional()
});

// Session progress schema
export const SessionProgressSchema = z.object({
  activitiesCompleted: z.array(z.string()).default([]),
  currentActivity: z.string().optional(),
  score: z.number().min(0).max(100).default(0),
  timeSpent: z.number().default(0), // in minutes
  engagementLevel: z.enum(["low", "medium", "high"]).optional(),
  hintsUsed: z.number().default(0)
});

// Learning session schema
export const SessionSchema = z.object({
  id: z.string().uuid(),
  childId: z.string().uuid(),
  lessonId: z.string().uuid(),
  subject: SubjectEnum,
  agentId: z.string(),
  
  // Session state
  status: z.enum(["active", "paused", "completed", "abandoned"]),
  startTime: z.string().datetime(),
  endTime: z.string().datetime().optional(),
  
  // Progress and content
  progress: SessionProgressSchema,
  messages: z.array(ChatMessageSchema).default([]),
  
  // Assessment
  finalScore: z.number().min(0).max(100).optional(),
  pointsEarned: z.number().min(0).default(0),
  badgesEarned: z.array(z.string()).default([]),
  
  // Metadata
  deviceInfo: z.object({
    userAgent: z.string().optional(),
    platform: z.string().optional(),
    screenSize: z.string().optional()
  }).optional()
}).merge(TimestampSchema);

// Start session request
export const StartSessionSchema = z.object({
  childId: z.string().uuid(),
  lessonId: z.string().uuid(),
  agentId: z.string()
});

// Send message request
export const SendMessageSchema = z.object({
  sessionId: z.string().uuid(),
  content: z.string().min(1),
  contentType: z.enum(["text", "audio"]).default("text"),
  audioData: z.string().optional() // base64 encoded
});

export type ChatMessage = z.infer<typeof ChatMessageSchema>;
export type SessionProgress = z.infer<typeof SessionProgressSchema>;
export type Session = z.infer<typeof SessionSchema>;
export type StartSessionRequest = z.infer<typeof StartSessionSchema>;
export type SendMessageRequest = z.infer<typeof SendMessageSchema>;
